File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
    int lcs(string& text1, string& text2, int i, int j, vector<vector<int>>& dp)
    {
        if(i < 0 or j < 0)
            return 0;
        
        if(dp[i][j] != -1)
            return dp[i][j];
        
        if(text1[i] == text2[j])
            return dp[i][j] = lcs(text1, text2, i-1, j-1, dp) + 1;
        
        return dp[i][j] = max(lcs(text1, text2, i, j-1, dp), lcs(text1, text2, i-1, j, dp));
    }
public:
    int longestCommonSubsequence(string text1, string text2) {
        
        int m = text1.size(), n = text2.size();
        vector<vector<int>>dp(m, vector<int>(n, -1));
        return lcs(text1, text2, m - 1, n - 1, dp);
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        
        m, n = len(text1), len(text2)
        
        dp = {}
        
        def lcs(i, j):

            if i == m or j == n:
                return 0
            
            if (i, j) in dp:
                return dp[i, j]
            
            if text1[i] == text2[j]:
                dp[i, j] =  1 + lcs(i + 1, j + 1)
            else:
                dp[i, j] = max(lcs(i, j + 1), lcs(i + 1, j)) 
                
            return dp[i, j]
                       
        return lcs(0, 0)
            
            
            
        
        